File: experiments/selenium-java/saucedemo-e2e/src/main/java/com/qaskills/saucedemo/pages/CheckoutPage.java
package com.qaskills.saucedemo.pages;

import org.openqa.selenium.By;
import org.openqa.selenium.WebDriver;

public class CheckoutPage extends BasePage {

    private static final By FIRST_NAME_INPUT = By.id("first-name");
    private static final By LAST_NAME_INPUT = By.id("last-name");
    private static final By POSTAL_CODE_INPUT = By.id("postal-code");
    private static final By CONTINUE_BUTTON = By.id("continue");
    private static final By FINISH_BUTTON = By.id("finish");
    private static final By CANCEL_BUTTON = By.id("cancel");
    private static final By SUBTOTAL_LABEL = By.cssSelector(".summary_subtotal_label");
    private static final By TAX_LABEL = By.cssSelector(".summary_tax_label");
    private static final By TOTAL_LABEL = By.cssSelector(".summary_total_label");
    private static final By COMPLETE_HEADER = By.cssSelector(".complete-header");

    public CheckoutPage(WebDriver driver) {
        super(driver);
    }

    public void waitForStepOne() {
        waitForUrlContains("checkout-step-one.html");
        wait.until(driver -> isDisplayed(FIRST_NAME_INPUT));
    }

    public CheckoutPage fillCustomerInfo(String firstName, String lastName, String postalCode) {
        type(FIRST_NAME_INPUT, firstName);
        type(LAST_NAME_INPUT, lastName);
        type(POSTAL_CODE_INPUT, postalCode);
        return this;
    }

    public CheckoutPage continueToOverview() {
        click(CONTINUE_BUTTON);
        return this;
    }

    public void waitForOverview() {
        waitForUrlContains("checkout-step-two.html");
        wait.until(driver -> isDisplayed(SUBTOTAL_LABEL) && isDisplayed(TAX_LABEL) && isDisplayed(TOTAL_LABEL));
    }

    public CheckoutPage completeOrder() {
        click(FINISH_BUTTON);
        return this;
    }

    public void waitForComplete() {
        waitForUrlContains("checkout-complete.html");
        wait.until(driver -> getText(COMPLETE_HEADER).contains("Thank you for your order!"));
    }

    public CartPage cancel() {
        click(CANCEL_BUTTON);
        return new CartPage(driver);
    }

    public boolean isProductVisible(String productName) {
        return isDisplayed(By.xpath("//*[contains(text(),'" + productName + "')]"));
    }

    public String getFirstNameError() {
        return getText(By.cssSelector("[data-test='error']"));
    }
}

File: experiments/cypress-e2e/saucedemo-e2e/cypress/pages/checkout.page.ts
export class CheckoutPage {
  get firstNameInput() {
    return cy.getByTest('firstName');
  }

  get lastNameInput() {
    return cy.getByTest('lastName');
  }

  get postalCodeInput() {
    return cy.getByTest('postalCode');
  }

  get continueButton() {
    return cy.getByTest('continue');
  }

  get finishButton() {
    return cy.getByTest('finish');
  }

  get cancelButton() {
    return cy.getByTest('cancel');
  }

  get subtotalLabel() {
    return cy.get('.summary_subtotal_label');
  }

  get taxLabel() {
    return cy.get('.summary_tax_label');
  }

  get totalLabel() {
    return cy.get('.summary_total_label');
  }

  fillCustomerInfo(firstName: string, lastName: string, postalCode: string) {
    this.firstNameInput.clear().type(firstName);
    this.lastNameInput.clear().type(lastName);
    this.postalCodeInput.clear().type(postalCode);
    return this;
  }

  continueToOverview() {
    this.continueButton.click();
    return this;
  }

  completeOrder() {
    this.finishButton.click();
    return this;
  }

  assertStepOne() {
    cy.url().should('include', 'checkout-step-one.html');
    this.firstNameInput.should('be.visible');
    return this;
  }

  assertOverview() {
    cy.url().should('include', 'checkout-step-two.html');
    this.subtotalLabel.should('be.visible');
    this.taxLabel.should('be.visible');
    this.totalLabel.should('be.visible');
    return this;
  }

  assertComplete() {
    cy.url().should('include', 'checkout-complete.html');
    cy.contains('Thank you for your order!').should('be.visible');
    return this;
  }
}

export const checkoutPage = new CheckoutPage();
